### ISSUE TITLE
Issues with pytest usage

### ISSUE THREAD

**Issue by MarcusJellinghaus** [MarcusJellinghaus/mcp-code-checker#36]
# Report by LLM that had issues

# Pytest Checker Tool Analysis

### **What I Submitted vs What I Received**

1. **Inconsistent Results Format**:
   ```python
   # Sometimes I got detailed failure info:
   "Pytest found issues that need attention: The following tests failed..."
   
   # Other times just summary:
   "Pytest check completed. Collected 416 tests in 0.99 seconds"
   
   # And sometimes:
   "Pytest check completed. Collected 394 tests in 1.06 seconds"
   ```

2. **Missing Failure Details**:
   ```python
   # I submitted:
   checker on p config:run_pytest_check(extra_args=["-k", "intellij"], verbosity=2)
   
   # Expected: Detailed failure information for intellij tests
   # Received: "Pytest check completed. Collected 394 tests in 1.06 seconds"
   # Missing: The actual failure details I needed to debug
   ```

3. **Inconsistent Test Counts**:
   - Sometimes 394 tests, sometimes 416 tests
   - No clear explanation for the variation

### **Tool Behavior Issues I Encountered**

1. **Silent Failures**: 
   ```python
   # This should have shown test failures but didn't:
   checker on p config:run_pytest_check(extra_args=["-k", "intellij_handler"], verbosity=2)
   # Result: Generic success message despite actual failures
   ```

2. **Timeout/No Response**:
   ```python
   # This completely failed:
   <error>No result received from client-side tool execution.</error>
   ```

3. **Inconsistent Detail Levels**:
   - Sometimes got full tracebacks
   - Sometimes got nothing despite `verbosity=2`

### **Implementation Issues (Most Likely)**

1. **Error Handling Problems**:
   - Tool appears to catch exceptions and return generic success messages
   - Should propagate actual pytest output even when tests fail
   - Missing proper error boundary handling

2. **Output Parsing Issues**:
   - Tool might be inconsistently parsing pytest's output
   - Different pytest exit codes might be handled differently
   - JSON parsing of pytest output could be failing silently

3. **Timeout Handling**:
   - Tool might have timeouts that aren't properly communicated
   - Large test suites might exceed time limits without proper feedback

### **Documentation Issues**

1. **Unclear Return Behavior**:
   ```python
   # Documentation should specify:
   - When will I get detailed failure info vs summary?
   - What does "verbosity" actually control?
   - How are different pytest exit codes handled?
   ```

2. **Missing Error Cases**:
   - No documentation on what happens when tests fail
   - No guidance on timeout behavior
   - No explanation of test count variations

### **Tool Implementation Improvements**

1. **Consistent Output Format**:
   ```python
   # Always return structured data:
   {
       "status": "failed" | "passed" | "error",
       "summary": "393 passed, 23 skipped, 9 failed",
       "details": "full pytest output",
       "failures": [...],  # parsed failure info
       "duration": "21.23 seconds"
   }
   ```

2. **Better Error Handling**:
   ```python
   # Should never silently fail:
   try:
       result = run_pytest(...)
       return parse_output(result)
   except Exception as e:
       return {"status": "error", "message": str(e), "raw_output": result.stdout}
   ```

3. **Timeout Management**:
   ```python
   # Clear timeout communication:
   if timeout_exceeded:
       return {"status": "timeout", "partial_output": output_so_far}
   ```

### **Documentation Improvements**

1. **Clear Behavior Specification**:
   ```markdown
   ## Return Behavior
   - `verbosity=0`: Summary only
   - `verbosity=1`: Summary + failure count
   - `verbosity=2`: Full failure details + tracebacks
   
   ## Error Handling
   - Test failures: Returns failure details
   - Tool errors: Returns error message
   - Timeouts: Returns timeout notification
   ```

2. **Example Usage**:
   ```python
   # Good examples of what to expect:
   result = run_pytest_check(extra_args=["-k", "failing_test"], verbosity=2)
   # Expected result format shown
   ```

## 🎯 **My Usage Issues**

1. **Over-reliance on Tool**: I kept calling the tool expecting different results without first checking if my fixes were actually correct
2. **Poor Error Interpretation**: When I got generic success messages, I should have questioned whether the tool was working correctly
3. **Insufficient Validation**: I should have manually run pytest to verify the tool's output

### **For Tool Implementation**
1. **Always return pytest's raw output** in addition to parsed summaries
2. **Implement proper timeout handling** with clear user communication
3. **Add debug mode** that shows exactly what pytest command was run
4. **Consistent error propagation** - never hide failures behind success messages

### **For Documentation**
1. **Add troubleshooting section** for when results seem inconsistent
2. **Specify exact output formats** for different scenarios
3. **Include examples** of both successful and failed test runs
4. **Document known limitations** and workarounds

## 📝 **Conclusion**

The main issue appears to be that the pytest checker tool has **implementation problems with error handling and output consistency**, combined with **insufficient documentation** about its actual behavior in edge cases. This led to a frustrating debugging experience where I couldn't trust the tool's output.

The primary problems were:
- Silent failures that returned success messages when tests actually failed
- Inconsistent output formats depending on internal tool state
- Missing timeout handling and communication
- Inadequate documentation of edge cases and error scenarios

These issues significantly impacted the debugging workflow and made it difficult to efficiently identify and fix the actual test problems.
